grift/src/GRIFT/Semantics/Expand.hs

```diff
@@ -1,53 +1,96 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero Public License for more details.
+
+You should have received a copy of the GNU Affero Public License
+along with GRIFT.  If not, see <https://www.gnu.org/licenses/>.
+-}
+
 {-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ExplicitNamespaces #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
 
 module GRIFT.Semantics.Expand where
 
 import Control.Lens ((^.))
-import Data.BitVector.Sized
-import Data.Parameterized
+import Data.Parameterized ( LeqProof(LeqProof), testLeq )
 import qualified Data.Sequence as Seq
 import Data.Sequence (Seq)
+import GHC.TypeLits ( type (<=) )
 
+import GRIFT.BitVector.BVApp
+import GRIFT.BitVector.BVFloatApp ( canonicalNaN32 )
 import GRIFT.Semantics
 import GRIFT.Semantics.Utils
-import GRIFT.Types
+import GRIFT.Types ( KnownRV, RVWidth, unSized )
 
 -- | Expand an 'AbbrevApp' into its unabbreviated version.
-expandAbbrevApp :: forall expr rv w . KnownRV rv => AbbrevApp expr rv w -> expr rv w
+expandAbbrevApp ::
+  forall rv w expr.
+  BVExpr (expr rv) =>
+  -- Num (expr rv 32) =>
+  KnownRV rv =>
+  AbbrevApp expr rv w -> expr rv w
 expandAbbrevApp (SafeGPRApp wRepr ridE) = iteE
-  (ridE `eqE` litBV 0)
-  (litBV (bitVector' wRepr (0 :: Integer)))
+  (ridE `eqE` bvInteger 0)
+  (bvInteger 0)
   (stateExpr (LocApp (GPRApp wRepr ridE)))
-expandAbbrevApp (ReadCSRApp _ csr) = cases
-  [ (csr `eqE` (litBV $ encodeCSR FFlags)
-    , let fcsr = rawReadCSR (litBV $ encodeCSR FCSR)
-          flags = extractE' (knownNat @5) (knownNat @0) fcsr
-      in zextE flags
-    )
-  , (csr `eqE` (litBV $ encodeCSR FRm)
-    , let fcsr = rawReadCSR (litBV $ encodeCSR FCSR)
-          rm = extractE' (knownNat @3) (knownNat @5) fcsr
-      in zextE rm
-    )
-  ]
-  (rawReadCSR csr)
-expandAbbrevApp (NanBox32App _ e) = (litBV (-1) :: expr rv 32) `concatE` e
+expandAbbrevApp (ReadCSRApp _ csr) =
+  -- NOTE: the following code z-extends a 5-bit value, thus requiring '5 <= w'.
+  -- However, adding this as a pre-condition is a bit of a nightmare, so let's see
+  -- if we can get away with a dynamic check until we have a better solution.
+  case testLeq (knownNat @5) (knownNat @w) of
+    Just LeqProof ->
+      withLeqTrans (knownNat @3) (knownNat @5) (knownNat @w) $
+      cases
+        [ (csr `eqE` bvExpr (encodeCSR FFlags)
+          , let fcsr = rawReadCSR (bvExpr $ encodeCSR FCSR)
+                flags = extractE' (knownNat @5) (knownNat @0) fcsr
+            in zextEOrId flags
+          )
+        , (csr `eqE` bvExpr (encodeCSR FRm)
+          , let fcsr = rawReadCSR (bvExpr $ encodeCSR FCSR)
+                rm = extractE' (knownNat @3) (knownNat @5) fcsr
+            in zextEOrId rm
+          )
+        ]
+        (rawReadCSR csr)
+    Nothing -> error "TODO"
```

Comment:
Can we add a better error message here? Also, it would be useful to have a `HasCallStack` constraint on `expandAbbrevApp` to provide a better error if this actually happens